### ISSUE TITLE
What is the recommended way to distribute requests across multiple threads?

### ISSUE THREAD

**Issue by zserik** [DataDog/glommio#237]
This is probably mostly a documentation thing... The first question that comes to my mind when thinking about building a server with TPC design and glommio is: how should I distribute requests across multiple threads (probably via channels)?

There are probably three different approaches to this:
1. Spread accepted TCP connections across threads/executors (this is possible using `glommio::net::AcceptedTcpStream`).
2. Manage all accepted TCP connections in one thread, but divide and spread individual tasks invoked by client requests across multiple threads.
3. Like the first approach, but combine it with "work-stealing", like `rayon`...

The first approach has the advantage that the latency is probably lower, because it doesn't need to perform much cross-thread communication, but it has the disadvantage that one connection could starve, impact or reduce performance of other connections "running" in the same thread. Example: A server accepts 8 connections and spreads them evenly across 4 threads with one executor per thread and one thread per core. All but two connections which unluckily end up in the same thread, are terminated (successful or unsuccessful doesn't matter) soon. If the remaining two connections both incur a huge load on the server, then only one core is utilized. This is bad.

The second approach has the advantage that it doesn't suffer from the same starvation problem that much (it only suffers from that problem if mostly data is transferred across the network via that thread, but only extremely small amouts of work can be offloaded to other threads). This approach has the disadvantage that the latency probably is higher, especially, if a huge amount of data is transferred over the network, and not much preprocessing is done on the server.

The third approach is probably the most appropriate one, for many generic workloads (althrough the other approaches are probably completely fine for short-lived connections and other special cases). It doesn't suffer from the starvation issue, as other executors would notice that they do nothing, and would try to steal the connections that is currently not processed (e.g. assume one connection is currently handled by the executor, it e.g. retrieves data from it, then other threads "steal" the currently waiting connection). Thus, this would repeatedly rebalance the connections across the threads as soon as one thread becomes idle. I think that glommio should support that use case somehow.

**Comment by glommer** [comment]
tl;dr: I am 100% onboard, but I don't think we should rush to a solution.


Connection balancing was by far one of the biggest issues in seastar applications, so I fully agree we need to have a good story here and I am very open to ideas.

One thing to keep in mind is that the receive path for thread-per-core applications is really tricky if you are chasing optimality: It doesn't matter what you do to balance the connections: the data still lives in a particular thread (which is the whole point behind much of the model) so you can process it in a lock-free manner, and now you have to route things to the owner thread.

Yes, having threads starved on the connection handling only makes things worse but the holy grail in thread per core is reducing cross-thread communication to zero. Scylla handled this with client cooperation (see https://github.com/scylladb/scylla/blob/master/docs/protocol-extensions.md "Intranode Sharding") but I believe we can have a better story here that is easier and more generic.

One area I want to experiment with -- I was talking to @bryandmc just yesterday about it - is that I believe we can use eBPF to distribute packets given a user-defined routing function into a couple of sockets that all have REUSEPORT and are on different executors. In a database setting, you likely will have different threads handling different keys and then your routing function can place the data into the right thread.

However, I also saw a lot of ideas proving counterintuitively bad (or counterintuitively good!) in this area, so I'd like those examples to come from real applications to the extent possible - even if we have to wait a bit to solidify the advices. As a matter of fact, I think I would be down to merge any proposed interface for this - including your thread stealing idea - under an experimental module so we can play with, report results, and compare approaches.

Likely different approaches will work better for different applications.

P.S.: have you considered joining our zulip chat? There is a lot of exciting conversations happening there

**Comment by bryandmc** [comment]
Work stealing is the approach used by all the other major rust async runtimes. That, in some ways, is the reason why I think there's massive advantage in not necessarily falling down that same hole. You could even implement a similar idea to work-stealing (except its more like, "work diverting" than stealing) using ebpf and REUSPORT. Though I agree with @glommer that having that be more user-defined would be more ideal if it can be accomplished. 

All that said, if there's a way to keep experiments in work-stealing or whatever in an experimental module that's an optional feature, I'd also be on board for that. Because there's not a ton of risk to just try stuff in that case. 

Those are my 2 cents

**Comment by qiujiangkun** [comment]
For those who need a work stealing approach, you can refer to my implementation.
https://github.com/qiujiangkun/async_executors/blob/dev/src/runtime/glommio_tp.rs
